dpctl/tensor/libtensor/include/kernels/linalg_functions/gemm.hpp

```diff
@@ -940,2165 +984,1092 @@ class gemm_k_krn;
+            exec_q, lhs_tp, rhs_tp, res_tp, single_batch_nelems, n, k, m,
+            batch_indexer, lhs_indexer, rhs_indexer, res_indexer,
+            {res_init_ev});
     }
+
+    return gemm_detail::_gemm_old_nm_impl<
+        lhsTy, rhsTy, resTy, BatchIndexerT, OuterInnerIndexerT,
+        OuterInnerIndexerT, OuterInnerIndexerT>(
+        exec_q, lhs_tp, rhs_tp, res_tp, single_batch_nelems, n, k, m,
+        batch_indexer, lhs_indexer, rhs_indexer, res_indexer, {res_init_ev});
 }
 
-template <typename T1, typename T2, typename T3> class gemm_tree_empty_krn;
+typedef sycl::event (*gemm_contig_impl_fn_ptr_t)(
+    sycl::queue &,
+    const char *, // lhs
+    const char *, // rhs
+    char *,       // res
+    size_t,       // n
+    size_t,       // k
+    size_t,       // m
+    std::vector<sycl::event> const &);
 
 template <typename lhsTy, typename rhsTy, typename resTy>
-sycl::event gemm_tree_impl(sycl::queue &exec_q,
-                           const char *lhs_cp,
-                           const char *rhs_cp,
-                           char *res_cp,
-                           size_t n,
-                           size_t k,
-                           size_t m,
-                           int inner_nd,
-                           int lhs_outer_nd,
-                           const ssize_t *lhs_outer_inner_shapes_strides,
-                           int rhs_outer_nd,
-                           const ssize_t *rhs_outer_inner_shapes_strides,
-                           int res_nd,
-                           const ssize_t *res_shapes_strides,
-                           std::vector<sycl::event> const &depends = {})
+sycl::event gemm_contig_impl(sycl::queue &exec_q,
+                             const char *lhs_cp,
+                             const char *rhs_cp,
+                             char *res_cp,
+                             size_t n,
+                             size_t k,
+                             size_t m,
+                             std::vector<sycl::event> const &depends = {})
 {
     const lhsTy *lhs_tp = reinterpret_cast<const lhsTy *>(lhs_cp);
     const rhsTy *rhs_tp = reinterpret_cast<const rhsTy *>(rhs_cp);
     resTy *res_tp = reinterpret_cast<resTy *>(res_cp);
 
-    if (k == 0) {
-        sycl::event gemm_no_reduction_ev =
-            exec_q.submit([&](sycl::handler &cgh) {
-                cgh.depends_on(depends);
+    using OuterInnerIndexerT = dpctl::tensor::offset_utils::NoOpIndexer;
+    constexpr OuterInnerIndexerT lhs_indexer{};
+    constexpr OuterInnerIndexerT rhs_indexer{};
+    constexpr OuterInnerIndexerT res_indexer{};
+
+    using BatchIndexerT = dpctl::tensor::offset_utils::ThreeZeroOffsets_Indexer;
+    constexpr BatchIndexerT batch_indexer{};
+
+    constexpr size_t single_batch_nelems = 1;
+
+    const size_t min_nm = std::min(n, m);
+    const size_t max_nm = std::max(n, m);
+    if (min_nm > 0 && (max_nm >= ((64 * 1024) / min_nm))) {
```

Comment:
Nit but `min_nm` is guaranteed to be greater than 0 by `py_dot`, which has an early exit path for `m == 0` or `n == 0`, since these arrays are empty.

Reply:
Ok, but since somebody can call gemm functions using kernel headers and not ensure this condition, it is best to leave this check as is.